REACT_STYLEGUIDE.MD

```diff
@@ -0,0 +1,192 @@
+
+Good:
+```
+<hello-component fname="person.fname" lname="person.lname" watch-depth="reference"></hello-component>
+```
+Bad:
+```
+<react-component name="HelloComponent" props="person" watch-depth="reference"/>
+```
+
+### Prefix ui_framework elements with kui, but not their file names.
+Good:
+```
+button.js:
+export function KuiButton({ onClick, isDisabled }) {
+  return <button className="kuiButton" onClick={onClick} isDisabled={isDisabled}/>
+};
+```
+Bad:
+```
+button.js:
+export class Button extends React.Component {
+  render() {
+    const { onClick, isDisabled } = this.props;
+    return <button className="kuiButton" onClick={onClick} isDisabled={isDisabled}/>
+  }
+}
+```
+The filenames leave it off because snake casing already increases file name length.
+
+### Action function names and prop function names
+
+Name action functions in the form of a strong verb and passed properties in the form of on<Subject><Change>. E.g:
+```
+<sort-button onClick={action.sort}/>
+<pagerButton onPageNext={action.turnToNextPage} />
+```
+
+### Avoid creating a function and passing that as a property, in render functions.
+Best (relies on [stage 2 proposal](https://github.com/tc39/proposal-class-public-fields)):
+```
+export class ClickCounter extends React.Component {
+  state = { clickCount: 0 };
+
+  // This syntax ensures `this` is bound within handleClick
+  onClick = () => {
+    this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+  }
+
+  render() {
+    return <button className="kuiButton" onClick={this.onClick} />
+  }
+}
+```
+Good:
+```
+export class ClickCounter extends React.Component {
+  constructor() {
+    this.state = { clickCount: 0 };
+    this.onClick = this.onClick.bind(this);
+  }
+
+  onClick() {
+    this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+  }
+
+  render() {
+    return <button className="kuiButton" onClick={this.onClick}/>
+  }
+}
+```
+
+Bad:
+```
+export class ClickCounter extends React.Component {
+  state = { clickCount: 0 };
+
+  onClick() {
+    this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+  }
+
+  render() {
+    return <button className="kuiButton" onClick={() => this.onClick()}/>
+  }
+}
+```
+Background: https://facebook.github.io/react/docs/handling-events.html
+There is also an eslint rule we should be able to turn on for this.
+
+### Never mutate state directly
+Good:
+```
+this.setState(prevState => { clickCount: prevState.clickCount + 1 });
+```
+Bad:
+```
+this.state.clickCount += 1;
+```
+
+### Prefer primitives over objects when storing in state.
+Good:
+```
+this.setState({
+  currentPage: 0,
+  selectedIds: []
+});
+```
+
+Discouraged:
+```
+this.setState({
+  pager: new Pager(),
+  selectedIds: new SelectedIds()
+});
+```
+
+## General Guidelines
+### Prefer pure functions when possible
```

Comment:
Awesome! I had a previous discussion in a PR about the merits of this: https://github.com/elastic/kibana/pull/11996#discussion_r118566279